> In MySQL, `BINARY` might not work for a primary key.
> 
> ```
>         experiments = self.store.search_experiments(filter_string="tag.KEY = 'value'")
> >       assert len(experiments) == 0
> E       AssertionError: assert 1 == 0
> E        +  where 1 = len([<Experiment: artifact_location='artifact_folder/1', experiment_id='1', lifecycle_stage='active', name='exp1', tags={'key': 'value'}>])
> ```
> 
> If it's challenging to fix this issue, we should document it as a limitation for MySQL. I can help.

what about remove "key" column from the primary-key ?
we have "experiment_id" as  primary-key so that it won't hurt performance too much.

--- a/mlflow/utils/search_utils.py
+++ b/mlflow/utils/search_utils.py
@@ -106,14 +106,22 @@ class SearchUtils:
             column = column.collate("Japanese_Bushu_Kakusu_100_CS_AS_KS_WS")
 
         # Use non-binary ahead of binary comparison for runtime performance
+        col = f"{column.class_.__tablename__}.{column.key}"
+        # Use non-binary ahead of binary comparison for runtime performance
         def case_sensitive_mysql_eq(value):
-            return sa.and_(column.__eq__(value), column.op("= BINARY", is_comparison=True)(value))
+            return sa.text(f"{col} = :value AND BINARY {col} = :value").bindparams(
+                sa.bindparam("value", value=value, unique=True)
+            )
 
         def case_sensitive_mysql_ne(value):
-            return sa.or_(column.__ne__(value), column.op("!= BINARY", is_comparison=True)(value))
+            return sa.text(f"{col} != :value OR BINARY {col} != :value").bindparams(
+                sa.bindparam("value", value=value, unique=True)
+            )
 
         def case_sensitive_mysql_like(value):
-            return sa.and_(column.like(value), column.op("LIKE BINARY", is_comparison=True)(value))
+            return sa.text(f"{col} LIKE :value AND BINARY {col} LIKE :value").bindparams(
+                sa.bindparam("value", value=value, unique=True)
+            )
 
         sql_filter_ops = {
             "=": case_sensitive_mysql_eq if dialect == MYSQL else column.__eq__,

--- a/mlflow/utils/search_utils.py
+++ b/mlflow/utils/search_utils.py
@@ -101,12 +101,13 @@ class SearchUtils:
     def get_sql_filter_ops(cls, column, operator, dialect):
         import sqlalchemy as sa
 
+        col = f"{column.class_.__tablename__}.{column.key}"
+
         # Use case-sensitive collation for MSSQL
         if dialect == MSSQL:
             column = column.collate("Japanese_Bushu_Kakusu_100_CS_AS_KS_WS")
 
         # Use non-binary ahead of binary comparison for runtime performance
-        col = f"{column.class_.__tablename__}.{column.key}"
         # Use non-binary ahead of binary comparison for runtime performance
         def case_sensitive_mysql_eq(value):
             return sa.text(f"{col} = :value AND BINARY {col} = :value").bindparams(

